### PR TITLE
Ra dspdc 1128 schema validation

### PR DESCRIPTION
**Pull request by raaidbroad**
Add json schema validation to our piece of the HCA pipeline. Between the additional Circe JSON encoding and the requests-scala library, I don't think this is very high performance, but it sure seems to work.

### CODE REVIEW

File: transformation/src/main/scala/org/broadinstitute/monster/hca/HcaPipelineBuilder.scala [hunk 1]
```diff
@@ -1,9 +1,14 @@
 package org.broadinstitute.monster.hca
 
+import cats.data.Validated
 import com.spotify.scio.ScioContext
 import com.spotify.scio.coders.Coder
 import com.spotify.scio.io.ClosedTap
 import com.spotify.scio.values.SCollection
+
+import scala.util.{Failure, Success}
```
Comment (danxmoran) [inline transformation/src/main/scala/org/broadinstitute/monster/hca/HcaPipelineBuilder.scala:9]: Nitpick: `scala.*` imports are typically grouped together at the bottom of the list. I've tried to automated it but scalafmt isn't smort enough yet.

File: transformation/src/main/scala/org/broadinstitute/monster/hca/HcaPipelineBuilder.scala [hunk 2]
```diff
@@ -275,6 +328,8 @@ object HcaPipelineBuilder extends PipelineBuilder[Args] {
     val metadataFiles = getReadableFiles(s"$inputPrefix/metadata/${entityType}/**.json", context)
     val metadataFilenameAndMsg = jsonToFilenameAndMsg(entityType)(metadataFiles)
 
+    validateJson(metadataFilenameAndMsg)
```
Comment (danxmoran) [inline transformation/src/main/scala/org/broadinstitute/monster/hca/HcaPipelineBuilder.scala:331]: This will probably work in practice, but it isn't _technically_ the right control-flow. Calling this as a `Unit` method here registers a side-branch of processing in the graph. It's completely disconnected from the steps below, so when the pipeline is sent to Dataflow it'll be run in parallel with processing, instead of in sequence.

The fix is to have `validateJson` return an `SCollection`, and use that output as inputs downstream.

File: transformation/src/main/scala/org/broadinstitute/monster/hca/HcaPipelineBuilder.scala [hunk 3]
```diff
@@ -256,6 +262,53 @@ object HcaPipelineBuilder extends PipelineBuilder[Args] {
   def encode(msg: Msg): String =
     upack.transform(msg, StringRenderer()).toString
 
+  def validateJson(filenamesAndMsg: SCollection[(String, Msg)]): Unit = {
```
Comment (danxmoran) [inline transformation/src/main/scala/org/broadinstitute/monster/hca/HcaPipelineBuilder.scala:265]:
```suggestion
  def validateJson(filenamesAndMsg: SCollection[(String, Msg)]): SCollection[(String, Msg)] = {
```
See comment below in the main method for explanation

File: transformation/src/main/scala/org/broadinstitute/monster/hca/HcaPipelineBuilder.scala [hunk 4]
```diff
@@ -256,6 +262,53 @@ object HcaPipelineBuilder extends PipelineBuilder[Args] {
...
+                    }
+                  }
+                }
+            }
+          }
+          case None => throw new Exception(s"No schema found at $url for file $filename")
+        }
+      }
+    }
+    ()
```
Comment (danxmoran) [inline transformation/src/main/scala/org/broadinstitute/monster/hca/HcaPipelineBuilder.scala:309]:
```suggestion
```

File: transformation/src/main/scala/org/broadinstitute/monster/hca/HcaPipelineBuilder.scala [hunk 5]
```diff
@@ -256,6 +262,53 @@ object HcaPipelineBuilder extends PipelineBuilder[Args] {
...
+              case Success(value) =>
+                // try to parse the actual data into a json format for validation
+                parse(encode(data)) match {
+                  case Left(_) =>
+                    throw new Exception(s"Unable to parse data into json for file $filename")
+                  // if everything is parsed/encoded/etc correctly, actually try to validate against schema here
+                  // if not valid, will return list of issues
+                  case Right(success) => {
+                    value.validate(success) match {
+                      case Validated.Valid(_) => ()
```
Comment (danxmoran) [inline transformation/src/main/scala/org/broadinstitute/monster/hca/HcaPipelineBuilder.scala:295]:
```suggestion
                      case Validated.Valid(_) => (filename, data)
```

File: transformation/src/test/scala/org/broadinstitute/monster/hca/HcaPipelineBuilderSpec.scala [hunk 6]
```diff
@@ -200,4 +211,83 @@ class HcaPipelineBuilderSpec extends AnyFlatSpec with Matchers {
...
+        |    "describedBy": "https://schema.humancellatlas.org/type/biomaterial/5.1.0/specimen_from_organism"
+        |}
+        |""".stripMargin
+    )
+
+    val exampleUrlAndFile = (exampleFileContent.read[String]("describedBy"), exampleFileContent)
+
+    the[Exception] thrownBy runWithContext { sc =>
+      HcaPipelineBuilder.validateJson(sc.parallelize(Seq(exampleUrlAndFile)))
+    } should have message "java.lang.Exception: Data does not conform to schema: NonEmptyList(#: required key [schema_type] not found)"
```
Comment (danxmoran) [inline transformation/src/test/scala/org/broadinstitute/monster/hca/HcaPipelineBuilderSpec.scala:291]: You could shorten this to be more flexible, in case there's any weird heuristics in the parser that might change the datatype used:
```suggestion
    } should include "Data does not conform to schema"
```

Comment (danxmoran) [reply transformation/src/test/scala/org/broadinstitute/monster/hca/HcaPipelineBuilderSpec.scala:291]: Oh, just noticed this is some magic helper for exceptions. Not sure if `include` works there.

File: transformation/src/main/scala/org/broadinstitute/monster/hca/HcaPipelineBuilder.scala [hunk 7]
```diff
@@ -256,6 +262,53 @@ object HcaPipelineBuilder extends PipelineBuilder[Args] {
   def encode(msg: Msg): String =
     upack.transform(msg, StringRenderer()).toString
 
+  def validateJson(filenamesAndMsg: SCollection[(String, Msg)]): Unit = {
+    // pull out the url for where the schema definition is for each file
+    val content = filenamesAndMsg.map {
+      case (filename, msg) => (msg.read[String]("describedBy"), (filename, msg))
+    }
+    // get the distinct urls (so as to minimize the number of get requests) and then get the schemas as strings
+    val schemas = content.map(_._1).distinct.map(url => (url, requests.get(url).text))
```
Comment (danxmoran) [inline transformation/src/main/scala/org/broadinstitute/monster/hca/HcaPipelineBuilder.scala:271]: OH, other thing to check: Try running this branch in Dataflow to make sure `requests` and the circe stuff are all serializable. If not, you'll get a runtime error.

File: transformation/src/main/scala/org/broadinstitute/monster/hca/HcaPipelineBuilder.scala (hunk 1 above)
Comment (raaidbroad) [reply transformation/src/main/scala/org/broadinstitute/monster/hca/HcaPipelineBuilder.scala:9]: tfw the automation does it wrong

File: transformation/src/main/scala/org/broadinstitute/monster/hca/HcaPipelineBuilder.scala (hunk 7 above)
Comment (raaidbroad) [reply transformation/src/main/scala/org/broadinstitute/monster/hca/HcaPipelineBuilder.scala:271]: ran it locally, it worked

File: transformation/src/test/scala/org/broadinstitute/monster/hca/HcaPipelineBuilderSpec.scala (hunk 6 above)
Comment (raaidbroad) [reply transformation/src/test/scala/org/broadinstitute/monster/hca/HcaPipelineBuilderSpec.scala:291]: Yeah I was following an example from the scala testing docs

File: transformation/src/main/scala/org/broadinstitute/monster/hca/HcaPipelineBuilder.scala (hunk 7 above)
Comment (danxmoran) [reply transformation/src/main/scala/org/broadinstitute/monster/hca/HcaPipelineBuilder.scala:271]: Can you link me to the dataflow job? So I can sanity-check the graph
